toolchain/semantics/semantics_handle_index.cpp

```diff
@@ -2,20 +2,65 @@
 // Exceptions. See /LICENSE for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-#include <utility>
-
 #include "toolchain/semantics/semantics_context.h"
+#include "toolchain/semantics/semantics_node.h"
+#include "toolchain/semantics/semantics_node_kind.h"
 
 namespace Carbon {
 
-auto SemanticsHandleIndexExpressionStart(SemanticsContext& context,
-                                         ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleIndexExpressionStart");
+auto SemanticsHandleIndexExpressionStart(SemanticsContext& /*context*/,
+                                         ParseTree::Node /*parse_node*/)
+    -> bool {
+  // Leave the expression on the stack for IndexExpression.
+  return true;
 }
 
 auto SemanticsHandleIndexExpression(SemanticsContext& context,
                                     ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleIndexExpression");
+  auto ind_node_id = context.node_stack().PopExpression();
+  auto name_node_id = context.node_stack().PopExpression();
+  auto name_node = context.semantics_ir().GetNode(name_node_id);
+
+  if (context.semantics_ir().GetNode(ind_node_id).kind() ==
+      SemanticsNodeKind::VarStorage) {
+    CARBON_DIAGNOSTIC(NondeterministicType, Error,
+                      "Type cannot be determined in compile time.");
+    context.emitter().Emit(parse_node, NondeterministicType);
+  } else if (name_node.kind() != SemanticsNodeKind::VarStorage) {
+    CARBON_DIAGNOSTIC(InvalidIndexExpression, Error,
+                      "Invalid index expression.");
+    context.emitter().Emit(parse_node, InvalidIndexExpression);
+  } else {
+    auto name_type = context.semantics_ir().GetNode(
+        context.semantics_ir().GetType(name_node.type_id()));
+    auto cast_ind_id = context.ImplicitAsRequired(
+        parse_node, ind_node_id,
+        context.CanonicalizeType(SemanticsNodeId::BuiltinIntegerType));
+    auto cast_ind_node = context.semantics_ir().GetNode(cast_ind_id);
+
+    if (cast_ind_node.type_id() != SemanticsTypeId::Error &&
+        name_type.kind() == SemanticsNodeKind::TupleType) {
+      auto ind_val = context.semantics_ir()
+                         .GetIntegerLiteral(cast_ind_node.GetAsIntegerLiteral())
+                         .getSExtValue();
+      auto type_block_id = name_type.GetAsTupleType();
+      auto type_block = context.semantics_ir().GetTypeBlock(type_block_id);
+      if (ind_val >= static_cast<int>(type_block.size())) {
+        CARBON_DIAGNOSTIC(OutOfBoundsAccess, Error,
+                          "Index: {0} is out of bound {1}.", int64_t, int64_t);
```

Comment:
How about something more like:

```
"Index `{0}` is past the end of `{1}`.", int64_t, std::string);
```

... sending the type as {1}

"past the end" makes it clearer that this is a greater-than reason for out-of-bounds than just "out of bound". Providing the type lets people see how it's being evaluated.

Reply:
It looks like you're passing the SemanticsNodeKind instead of the type. What I meant was a std::string if the type, which you should be able to get with `semantics_ir().StringifyType(name_node.type_id())`